### ISSUE TITLE
Database has a global failure DOMException: Maximum IPC message size exceeded.

### ISSUE THREAD

**Issue by DarrenPBrown** [pouchdb/pouchdb#8001]
Hello,

We are using PouchDB 7.1.1 on Android wrapped with Cordova. The database in use is indexed db.
Our app allows attachments to be saved to the local databases through PouchDB.
At various points in the app where reading or writing of the attachment to the database takes place we sometimes get the error **Database has a global failure DOMException: Maximum IPC message size exceeded.** the error is only reported to the console through. We do use .catch to error trap any PouchDB errors but this error does not seem to be received by our error trapping code.

I can manage to reproduce the issue by adding 6 to 8, 12mb attachments to the local database and then using functionality that acts upon these attachments i.e. deleting and re-adding one of them, or cycling through them with the map reduce to upload to our SAP system using OData calls.

Is there anything we can do to prevent this error occurring?

regards,

Darren.

**Comment by karlwestin** [comment]
these are very large attachments and it's possible you are hitting some internal indexeddb limit https://chromium.googlesource.com/chromium/src.git/+/master/content/browser/indexed_db/indexed_db_database.cc#1083

**Comment by DarrenPBrown** [comment]
Is there anything we can get from PouchDB that would inform us that the attachments are too large or there are too many.

We could place limits on attachment size and number but it would be best if these were based on values that are not going to cause errors.

**Comment by DarrenPBrown** [comment]
I'll do some testing around the content in the link below to see if the limit of 128Mb is the issue. If it's that value consistently then we can check the total used across all attachments for the work order (something that attachments are associated to).

[indexeddb.mojom](https://chromium.googlesource.com/chromium/src/third_party/+/master/blink/public/mojom/indexeddb/indexeddb.mojom?autodive=0%2F%2F%2F%2F)

**Comment by jeremy21212121** [comment]
You are running up against a limitation of the browser engine.

Serialized message passing is the only way workers, WASM modules, etc can communicate with the main JS thread. So all your reads/writes are bottlenecked by the serialize, pass, parse process, and you will get an error if the serialized data is over ~100-128mb. It really is quite inefficient for data-heavy stuff compared to shared memory. I believe some sort of shared memory is in the works for WASM, dunno about workers though 

The browser really can't handle much more than 100mb serializing/parsing/message passing in one go. I ran into this problem with a desktop web app on a contract about a year ago. All you can do is take care to read/write in less than 100, 128mb increments.

[This article](https://developers.redhat.com/blog/2014/05/20/communicating-large-objects-with-web-workers-in-javascript/)  talks about various approaches for dealing with this issue.

There is also a [relevant SO post](https://stackoverflow.com/questions/52717593/maximum-ipc-message-size-exceeded) here.

**Comment by klues** [comment]
I also ran into the `Maximum IPC message size exceeded` error in current Chrome (`Version 80.0.3987.149`) on a [bulkDocs()](https://pouchdb.com/guides/bulk-operations.html) operation with huge amout of data (about 140MB).

I solved it by saving data in smaller chunks like this:
```
function bulkSave (documentList) {
    let maxDocsSaveAtOnce = 50; //depends on how big your documents are
    let chunks = [];
    if (documentList.length > maxDocsSaveAtOnce) {
        chunks = util.splitInChunks(documentList, maxDocsSaveAtOnce);
    } else {
        chunks = [documentList];
    }
    function saveChunksSequentially(chunks) {
        return pouchDb.bulkDocs(chunks.shift()).then(() => {
            if (chunks.length > 0) {
                return saveChunksSequentially(chunks);
            } else {
                return Promise.resolve();
            }
        })
    }
    return saveChunksSequentially(chunks);
}
```
The method `util.splitInChunks` is something like one of [these suggestions on Stackoverflow](https://stackoverflow.com/questions/8495687/split-array-into-chunks).

For me this solved the `Maximum IPC message size exceeded` error, however i still got `indexed_db_went_bad` errors in Chrome. So it seems I'm hitting some other limits. However for Firefox (`Version 74.0`) it's now working without errors.

Suggestion for pouchDb developers: maybe it makes sense to implement some kind of chunk-splitting within the library if bulkDocs() gets too much data.

**Comment by DarrenPBrown** [comment]
Thank you to all for replying. With your assistance I have resolved the problem. There were two main culprits. 

1, Attachments were being saved as part of the document data in a property called attachment so when the document was being returned so was the attachment. I have changed that by saving the attachment data to the property called _attachments, see [Save an inline attachment](https://pouchdb.com/api.html#save_attachment).

2, The app used . query temporary query syntax so was slow. I have changed this to use persistent queries, and to only return the doc and to then use a get to return the doc again but with the attachment data using a get with options.attachments property set to true. See [Temporary and Persisent Queries](https://pouchdb.com/guides/queries.html) and [Fetch a document](https://pouchdb.com/api.html#fetch_document).

The message "Maxiumum IPC message size exceeded" no longer appears.

The app now handles attachments up to 70Mb in size and and I guess would probably handle more - I will put some restrictions on this though.

thanks, Darren.
